### ISSUE TITLE
TensorFlow Lite (2.3.0) Converter Quantization Incompatibility for Coral Edge TPU Compiler

### ISSUE THREAD

**Issue by goodwilj** [tensorflow/tensorflow#43275]
**System information**
- OS Platform and Distribution: Linux Ubuntu 20.04 LTS with Docker
- TensorFlow installed from:  pulled TensorFlow Docker image (version 19.03.12)
- TensorFlow version (or github SHA if from source): v2.3.0 CPU
- Coral Edge TPU Compiler: version 14.1.317412892

**Command used to run the converter or code if you’re using the Python API**
If possible, please share a link to Colab/Jupyter/any notebook.

```
import tensorflow as tf 
import numpy as np
from tensorflow import keras

keras_model = tf.keras.models.load_model("keras_model.h5")
keras_model.input.set_shape((1,)+keras_model.input.shape[1:])
keras_model.compile(optimizer="Adam", loss="mse", metrics=["mae"])
converter = tf.lite.TFLiteConverter.from_keras_model(keras_model)
#input_arrays=["x"], output_arrays=["Identity"],input_shapes={'x':[1,32,32,1]})


def representative_dataset_gen():
  for _ in range(10):
    input_array = np.random.random((1,32,32,1))
    input_array = np.array(input_array,dtype=np.float32)
    yield [input_array]

converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.representative_dataset = representative_dataset_gen
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
converter.inference_input_type = tf.uint8  # or tf.uint8
converter.inference_output_type = tf.uint8  # or tf.uint8
tflite_model = converter.convert()

with open('keras_model_2-3-0.tflite', 'wb') as f:
  f.write(tflite_model)
```
Note: I am using garbage random data to do the post-training quantization.

**The output from the converter invocation**

```
2020-09-16 18:08:33.821326: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'libcudart.so.10.1'; dlerror: libcudart.so.10.1: cannot open shared object file: No such file or directory
2020-09-16 18:08:33.825281: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
2020-09-16 18:08:38.269950: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory
2020-09-16 18:08:38.270738: W tensorflow/stream_executor/cuda/cuda_driver.cc:312] failed call to cuInit: UNKNOWN ERROR (303)
2020-09-16 18:08:38.271532: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (e386413cb5cf): /proc/driver/nvidia/version does not exist
2020-09-16 18:08:38.278651: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN)to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2020-09-16 18:08:38.337624: I tensorflow/core/platform/profile_utils/cpu_utils.cc:104] CPU Frequency: 1799995000 Hz
2020-09-16 18:08:38.345643: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x4714550 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2020-09-16 18:08:38.345866: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
WARNING:tensorflow:No training configuration found in the save file, so the model was *not* compiled. Compile it manually.
WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow/python/training/tracking/tracking.py:111: Model.state_updates (from tensorflow.python.keras.engine.training) is deprecated and will be removed in a future version.
Instructions for updating:
This property should not be used in TensorFlow 2.0, as updates are applied automatically.
2020-09-16 18:09:01.424650: W tensorflow/python/util/util.cc:348] Sets are not currently considered sequences, but this may change in the future, so consider avoiding using them.
WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow/python/training/tracking/tracking.py:111: Layer.updates (from tensorflow.python.keras.engine.base_layer) is deprecated and will be removed in a future version.
Instructions for updating:
This property should not be used in TensorFlow 2.0, as updates are applied automatically.
2020-09-16 18:09:28.840811: I tensorflow/core/grappler/devices.cc:69] Number of eligible GPUs (core count >= 8, compute capability >= 0.0): 0
2020-09-16 18:09:28.847221: I tensorflow/core/grappler/clusters/single_machine.cc:356] Starting new session
2020-09-16 18:09:28.910112: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:816] Optimization results for grappler item: graph_to_optimize
2020-09-16 18:09:28.910190: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:818]   function_optimizer: function_optimizer did nothing. time = 1.162ms.
2020-09-16 18:09:28.910203: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:818]   function_optimizer: function_optimizer did nothing. time = 0.004ms.
2020-09-16 18:09:31.373431: W tensorflow/compiler/mlir/lite/python/tf_tfl_flatbuffer_helpers.cc:313] Ignored output_format.
2020-09-16 18:09:31.373507: W tensorflow/compiler/mlir/lite/python/tf_tfl_flatbuffer_helpers.cc:316] Ignored drop_control_dependency.

Attached is the Keras model and tflite model below:
[models.zip](https://github.com/tensorflow/tensorflow/files/5234095/models.zip)

```

**Failure details**
I am attempting to use this post-training quantized TensorFlow Lite model for input to the Coral Edge TPU compiler such that it can run on the Edge TPU.  The Edge TPU compiler fails to compile the quantized TensorFlow Lite model.  I reached to the Coral Edge TPU team to examine why the the TensorFlow Lite model will not compile.  According to their team, "The compiler rejects the model on purpose because there are some mis matching in quantization parameter which could cause bad prediction:

![image](https://user-images.githubusercontent.com/19961323/93378124-fa01f000-f829-11ea-8fd8-2a88204dd4ac.png)

The are 2 quantized op going to that same Concat layer where one has this:
`scale: 0.048531219363212585 zero_point: 103 num_fxp_values: 256`
and the other one:

`scale: 0.033884394913911819 zero_point: 120 num_fxp_values: 256`
This seems to me like a conversion issue, I suggest reaching out to the tensorflow team for a more appropriate solution."

Why are there 2 quantized operations with different quantization parameters going to the same concatenation layer?  Is this an error with the TensorFlow Lite converter or an error with the Edge TPU compiler?

**Comment by amahendrakar** [comment]
@goodwilj,
In order to expedite the trouble-shooting process, could you please provide the code to generate the `keras_model.h5` file as well as the model (i.e. the `keras_model.h5` file) itself. Thanks!

**Comment by goodwilj** [comment]
Thanks for the reply! Attached below is the code to generate the keras_model.h5 and the model itself.

[fc_densenet.zip](https://github.com/tensorflow/tensorflow/files/5239226/fc_densenet.zip)

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with TF v2.3 and TF-nightly. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/e8d52cf7d93a47563b8f604eaa9ff446/43275.ipynb). Thanks!

**Comment by goodwilj** [comment]
Thanks! @amahendrakar

**Comment by thaink** [comment]
Reassigned to MOT team.

**Comment by magneter** [comment]
Help.
The exactly same problem occured to me. And I have tried many versions to avoid it ,but it did not worked.
How could I solve the problem as soon ? Thanks

**Comment by saikumarchalla** [comment]
@goodwilj I treid to reproduce the issue but facing different error in TF 2.5. Please find the gist [here](https://colab.research.google.com/gist/saikumarchalla/837dde655958409925695c086878f8f8/43275.ipynb#scrollTo=l7rcScsd21FX).Thanks!

**Comment by goodwilj** [comment]
@saikumarchalla This actually appears to be the exact same error as detailed in the original post, so I believe you have reproduced it.  According to the Edge TPU team, the crux of the problem is that there are two quantized operations with different scales/zero-points going to the same concatenation layer:

"The are 2 quantized op going to that same Concat layer where one has this:
`scale: 0.048531219363212585 zero_point: 103 num_fxp_values: 256`
and the other one:
`scale: 0.033884394913911819 zero_point: 120 num_fxp_values: 256` "

**Comment by mohantym** [comment]
Hi @goodwilj ! This issue is getting resolved in the [2.8](https://colab.sandbox.google.com/gist/mohantym/5579c86990354a36bc2a52547dbc5ff6/untitled67.ipynb) version . Can we move this issue to closed status now?

**Comment by goodwilj** [comment]
@mohantym Sounds good! Thank you for the update. We can move this to closed status.
